tests/nn/data_parallel/test_fsdp_freezing_weights.py

```diff
@@ -34,38 +35,41 @@ def __init__(self):
             nn.Flatten(),
         )
         self.head = nn.Linear(64, 10)
+        if with_fsdp and freeze_after_wrap_fsdp:
+            self.trunk = FSDP(self.trunk)
+            self.head = FSDP(self.head)
 
     def forward(self, x):
         return self.head(self.trunk(x))
 
 
 class NestedTrunkModel(nn.Module):
-    def __init__(self, with_fsdp):
+    def __init__(self, with_fsdp, freeze_after_wrap_fsdp):
         super().__init__()
-        self.trunk = nn.Sequential(self._create_block(3, 64, with_fsdp), self._create_block(64, 64, with_fsdp),)
+        self.trunk = nn.Sequential(
+            self._create_block(3, 64, with_fsdp, freeze_after_wrap_fsdp),
+            self._create_block(64, 64, with_fsdp, freeze_after_wrap_fsdp),
+        )
         self.head = nn.Sequential(nn.AdaptiveAvgPool2d(output_size=(1, 1)), nn.Flatten(), nn.Linear(64, 10),)
-        if with_fsdp:
+        if with_fsdp and freeze_after_wrap_fsdp:
             self.trunk = FSDP(self.trunk)
             self.head = FSDP(self.head)
 
     def forward(self, x):
         return self.head(self.trunk(x))
 
-    def _create_block(self, in_channels, out_channels, with_fsdp):
+    def _create_block(self, in_channels, out_channels, with_fsdp, freeze_after_wrap_fsdp):
         block = nn.Sequential(nn.Conv2d(in_channels, out_channels, kernel_size=3), nn.ReLU(inplace=True),)
-        if with_fsdp:
+        if with_fsdp and freeze_after_wrap_fsdp:
             block = FSDP(block)
```

Comment:
I think this line end up being different between frozen before and frozen after. I suggest we do the following:

```
model:
  def __init__:
    ...
    if with_fsdp and freeze_after_wrap_fsdp:
       self.fsdp_wrap()

  def fsdp_wrap():
    do some wrapping


... later in the test ...

  if not freeze_after_wrap_fsdp:
    model.fsdp_wrap()
  model = fsdp(model)
```

I think this way, we will ensure the wrapping happens in the same function in the same way in both cases. What do you think?

Reply:
sounds good!